### ISSUE TITLE
[BUG] Search doesn't work for 'İ' and 'ı'

### ISSUE THREAD

**Issue by faydin** [TiddlyWiki/TiddlyWiki5#4905]
**Describe the bug**
Searching for 'i' doesn't find 'İ' (the capital of the letter 'i' in Turkish) and vice versa. The same for 'I' and 'ı'.
Fixing this wouldn't effect the other languages.
So, in Turkish:
- i and İ are the same letters
- ı and I are the same letters

**To Reproduce**
1. Create a tiddler with 'İİİİİ' in it. (those are capital 'i's)
2. Search for 'iiiii'.
3. It will not find the tiddler.

**Expected behavior**
i and İ should be considered as the same letters.
ı and I should be considered as the same letters.

**Desktop (please complete the following information):**
 - OS: Linux
 - Browser: Firefox
 - Version: 5.1.23-prerelease

**Comment by Jermolene** [comment]
Hi @faydin thanks for the report. Under the hood, TiddlyWiki uses JavaScript regular expressions for searching. Regular expression search primitives offer support for case insensitivity, but it only works for basic English letters.

We do already have one component of a fix: there's a table of "transliterations" that converts accented characters into the closest equivalent basic English letters:

https://github.com/Jermolene/TiddlyWiki5/blob/master/core/modules/utils/transliterate.js

The mapping table does already include Turkish characters. So, to fix this we'd need to transliterate the search string and the text of each tiddler when it is searched. I think the fix would benefit most non-English languages, eg. it would allow searching "e" and retrieving "é".

**Comment by faydin** [comment]
So, will it be easy to implement?

**Comment by Jermolene** [comment]
Hi @faydin 

> So, will it be easy to implement?

Sadly, it isn't particularly straightforward because it is a modification to a key core function. For example, the transliteration process has significant performance implications, so we'd need to explore ways to retain the existing behaviour for users who don't need the extra processing.

**Comment by rmunn** [comment]
@Jermolene [Turkish is particularly complicated](http://www.i18nguy.com/unicode/turkish-i18n.html) to get case-insensitivity right for, for the reasons mentioned in the linked article. A lot of code out there, including Javascript regular expressions, assumes that the uppercase of i will be I and the lowercase of I will be i, but that's not true in Turkish, and when those assumptions break, lots of things go wrong.

I think it might be easy-ish to implement by coding the search to use [localeCompare](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/localeCompare), but that will probably be a performance degradation over regular expressions so it might be worth adding a TIddlyWiki option, defaulting to off, to use localeCompare instead of regexps. Something like "Use locale-aware text search?" And a tooltip or one-paragraph explanation saying "For some countries, e.g. Turkey, the default search gets it wrong. Locale-aware text search is slower but will get local capitalization rules correct in all cases." Note that I haven't tested this (yet) so my expectations of performance penalties are simply because I believe regexps are highly optimized in Javascript engines. But it might be that localeCompare will end up calling ICU functions, which are also quite well-optimized, to do the actual comparison, so using localeCompare might allow us to implement this optional feature at not too high a performance cost.

**Comment by Jermolene** [comment]
Thanks @rmunn. It's annoying that there doesn't appear to be a "localeIndexOf()" function.

Another factor to be considered is that we might want to have the option of using a proper JavaScript search engine for searching, so that we can do a better job of ranking results, and fuzzy matches etc. I've done some experimentation with http://elasticlunr.com/ in the past, but only got as far as discovering that the indexes that it creates are super-huge.

**Comment by rmunn** [comment]
@Jermolene https://github.com/arty-name/locale-index-of might work: it's got zero dependencies (it uses [Intl.Collator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Collator) which should be in every browser, including IE 11 if we still care about IE). The license is [ISC](https://en.wikipedia.org/wiki/ISC_license), which I had to look up but it's essentially the same as the MIT license with slightly less legalese. So there should be no problems, legal or technical, with incorporating it into TiddlyWiki.

**Comment by Jermolene** [comment]
Hi @rmunn 

> @Jermolene https://github.com/arty-name/locale-index-of might work: it's got zero dependencies (it uses [Intl.Collator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Collator) which should be in every browser, including IE 11 if we still care about IE). The license is [ISC](https://en.wikipedia.org/wiki/ISC_license), which I had to look up but it's essentially the same as the MIT license with slightly less legalese. So there should be no problems, legal or technical, with incorporating it into TiddlyWiki.

Amazing, if only everything one wished for turned out to already exist! But I notice that the [inner loop](https://github.com/arty-name/locale-index-of/blob/3988c568ef39c99e6392735c41dab6e3317449dd/index.js#L35-L40) is as brute force as one might fear:

```js
  for (var index = 0; index <= stringLength - substringLength; index += 1) {
    var potentialMatch = string.substring(index, index + substringLength);
    if (collator.compare(potentialMatch, substring) === 0) {
      return index;
    }
  }
```

There are a bunch of fascinating optimisations that can be done for plain text searching (eg if one doesn't find the search string of length `n` at position `p`, then as long as the search string doesn't contain repetition, one doesn't need to search again until `p+n`). I can see that the fuzziness here might make such things hard to implement, but it does seem a shame to have to settle for such poor performance.

**Comment by jms19** [comment]
I'd love search to be case and accent insensitive. It certainly isn't as of 5.2.1 .

**Comment by ght** [comment]
The way Firefox works around the Turkish ı/İ problem with case insensitive search for Ctrl+F is much simpler: they just fold all four letters (iIıİ) together and each matches each of the others (see [#812837 in their bugzilla](https://bugzilla.mozilla.org/show_bug.cgi?id=812837#c5)).

That should be simple to reproduce, by replacing each of those letters in a search term with *all* of them. So instead of `/tiddly/i` the search regex should be `/t[iıİ]ddly/i`.

In other words:
```js
RegExp(originalRegexp.source.replace(/[iıİ]/g, '[iıİ]'));
```

**Comment by pmario** [comment]
@ght That's tricky and as you wrote, it should be simple to implement.

**Comment by Jermolene** [comment]
Hi @ght 

> That should be simple to reproduce, by replacing each of those letters in a search term with all of them.

Yes, that technique seems to be a common solution, and I think it could be made to work for us. We already have a table of transliteration pairs which could be turned inside out in order to do the reverse mappings.

As others have pointed out, using an off the shelf full text search library may well be the best way to go so that we get other desirable features like stemming and stop words.

**Comment by pmario** [comment]
>As others have pointed out, using an off the shelf full text search library may well be the best way to go so that we get other desirable features like stemming and stop words.

The biggest problem I see with 3rd party libs is, that they are all huge in file size. So I personally would like to have a small working solution in the core and plugins for other search mechanisms. ... 

IMO the crux is, that all browsers offer search functions in the browser, but they don't offer a JS API to use them :/

**Comment by xcazin** [comment]
> The biggest problem I see with 3rd party libs is, that they are all huge in file size. So I personally would like to have a small working solution in the core and plugins for other search mechanisms. ...
> 
> IMO the crux is, that all browsers offer search functions in the browser, but they don't offer a JS API to use them :/

For years, `Intl.Collator`, the JavaScript implementation for multilingual collation, was not recognised by every browser. Now, I think we can reckon from this table https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Collator/Collator#browser_compatibility that full compatibility has been achieved. In that very reference page, one can try:

```
console.log(new Intl.Collator('tr', { sensitivity: 'base' } ).compare("İİİİİ", "iiiii")); // 0
console.log(new Intl.Collator('tr', { sensitivity: 'base' } ).compare("IIIII", "ııııı")); // 0
console.log(new Intl.Collator('de', { sensitivity: 'base' } ).compare("Für große Bösewichte", "fur grosse Bosewichte")); // 0
console.log(new Intl.Collator('fr', { sensitivity: 'base' } ).compare("L'Être et le néant", "l'etre et le neant")); // 0
console.log(new Intl.Collator('fr', { sensitivity: 'base' } ).compare("Cœur à défendre", "Coeur a defendre")); // 0
```

Obviously, I am mostly familiar with European languages, but it looks like they considered a lot of character sorting schemes. So I believe that users who want to search non English-only wikis could also see `Intl.Collator` as a valuable addition for searching and sorting features in TiddlyWiki. 

Eventhough there may be performance issues compared to transliteration, users of my knowledge bases at work would be more than happy to install a slowish plugin that allowed them to search non-English wikis with non-accented strings.

**Comment by Jermolene** [comment]
Thanks @xcazin. The frustration is that the `compare()` method provided by the collator object isn't quite what we need to implement a search operation. The comparison operation allows us to compare two strings in the same way that is needed for sorting, but there is no provision for a matching `indexOf()` method that can determine if one string occurs within another.

**Comment by Leilei332** [comment]
> The way Firefox works around the Turkish ı/İ problem with case insensitive search for Ctrl+F is much simpler: they just fold all four letters (iIıİ) together and each matches each of the others (see [#812837 in their bugzilla](https://bugzilla.mozilla.org/show_bug.cgi?id=812837#c5)).
> 
> 
> 
> That should be simple to reproduce, by replacing each of those letters in a search term with *all* of them. So instead of `/tiddly/i` the search regex should be `/t[iıİ]ddly/i`.
> 
> 
> 
> In other words:
> 
> ```js
> 
> RegExp(originalRegexp.source.replace(/[iıİ]/g, '[iıİ]'));
> 
> ```

I think this one is feasible. I wonder whether we can create a plugin to hook the tiddlywiki's regexp search mechanism, or make the search mechanism a bit pluggable.
